### PR TITLE
Getting googletest code can be compiled with -Wundef option

### PR DESCRIPTION
**Pull request by lygstate**
Signed-off-by: Yonggang Luo <luoyonggang@gmail.com>

### CODE REVIEW

**Review by tigthor — APPROVED** [review]
I have reviewed this, this is a good option

**Comment by lygstate** [comment]
The windows CI also fixed by this MR

**Comment by dinord** [comment]
1. This contains changes beyond the minimal code required to enable -Wundef (e.g. edits to build files, adding underscores to struct fields, etc.). Could we reduce it to just change the spurious macros?
2. Since we use the pattern `#if defined(__FOO) && __FOO` for external macros, could we use it for googletest macros as well? Instead of attempting to define them to 0.

**Comment by lygstate** [comment]
> 1. This contains changes beyond the minimal code required to enable -Wundef (e.g. edits to build files, adding underscores to struct fields, etc.). Could we reduce it to just change the spurious macros?
> 2. Since we use the pattern `#if defined(__FOO) && __FOO` for external macros, could we use it for googletest macros as well? Instead of attempting to define them to 0.

define them to 0 is better than `#if defined(__FOO) && __FOO` because we can use `-Wundef` to detect if the invalid macro are used

**Comment by lygstate** [comment]
@thughes Please take a look at this MR

File: googletest/include/gtest/internal/gtest-port-arch.h [hunk 1]
```diff
@@ -115,4 +115,100 @@
 #define GTEST_OS_QURT 1
 #endif  // __CYGWIN__
 
+// Defines macros that will be used but not defined to 0
```
Comment (krishnacx) [inline googletest/include/gtest/internal/gtest-port-arch.h:118]: Defines -> Define

File: googletest/include/gtest/internal/gtest-port.h [hunk 2]
```diff
@@ -648,6 +656,81 @@ typedef struct _RTL_CRITICAL_SECTION GTEST_CRITICAL_SECTION;
...
+#ifndef GTEST_DONT_DEFINE_EXPECT_FALSE
+#define GTEST_DONT_DEFINE_EXPECT_FALSE 0
+#endif
+#ifndef GTEST_DONT_DEFINE_TEST_F
+#define GTEST_DONT_DEFINE_TEST_F 0
+#endif
+#ifndef GTEST_DONT_DEFINE_TEST
+#define GTEST_DONT_DEFINE_TEST 0
+#endif
+
```
Comment (krishnacx) [inline googletest/include/gtest/internal/gtest-port.h:726]: Empty line can be ommitted.

**Comment by lygstate** [comment]
cc @dinord

**Comment by thughes** [comment]
This should now be fixed as of commit d92a270d2d9d6bb7ed3bf7dc3f698028550a65ef.

**Comment by lygstate** [comment]
@thughes There is an issue, the bazel didn't guard -Wundef option
      # Enable Warn if an undefined identifier is evaluated in an #if directive. Such identifiers are replaced with zero.
        "//:msvc_compiler": [],
        "//conditions:default": ["-Wundef"],

**Comment by thughes** [comment]
@lygstate Should be fixed in commit 678c1c73de2cf004d58f5dbbef299083997d924a and a798c2f10200b6293a5cc236b5f41b26c1ae7378.

**Pull request closed without merging**